### PR TITLE
Add ability to have a mirror postgres DB

### PR DESCRIPTION
**Pull request by darkruby501**
To make migrations easier and safer, this PR adds the ability to configure a "mirror" database (really a second destination) so the analytics will stream to two places.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1202571304226634) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by jpaddison3** [comment]
Like, you'll need to create a new DB, restore an old copy to it, and then start streaming to it, but then there'd be a gap in the data. Alternatively, you could create a new DB, start streaming to it, then copy a newer backup into it, and then maybe there's a unique index and you can do a `ON CONFLICT IGNORE`?

File: packages/lesswrong/server/analytics/postgresConnection.ts [hunk 1]
```diff
@@ -7,29 +7,40 @@ import { forumTypeSetting } from "../../lib/instanceSettings";
 export const pgPromiseLib = pgp({});
 
 export const connectionStringSetting = new DatabaseServerSetting<string | null>("analytics.connectionString", null);
```
Comment (jpaddison3) [inline packages/lesswrong/server/analytics/postgresConnection.ts:9]: I'd strongly consider making this a `<string | string[] | null>` setting

**Comment by jpaddison3** [comment]
I'd personally use transactions for this, but I understand if you want to do this.

**Comment by darkruby501** [comment]
>Can you write more about how this works? How do you get from today to where you have a redundant DB

>Like, you'll need to create a new DB, restore an old copy to it, and then start streaming to it, but then there'd be a gap in the data. Alternatively, you could create a new DB, start streaming to it, then copy a newer backup into it, and then maybe there's a unique index and you can do a ON CONFLICT IGNORE?

I was imagining doing something like the latter. Haven't completed the process yet, but adding a mirror seemed like a good thing to do before beginning a migration (we're going to go from self-hosted Postgres to RDS). We've got an RDS instance now and I've tested streaming to it from local instance. Wanted to get it in prod so I'm getting more data to play with.

>I'd strongly consider making this a <string | string[] | null> setting
I considered this but (1) seems unlikely we'll ever want more than one mirror. (2) Doing a list introduces not immediately obvious significance to the ordering in the list, since only one of your DBs is going to be your read-source.

>I'd personally use transactions for this, but I understand if you want to do this.
I'm not sure what exactly how this means or how you'd do this, so quite plausibly better!

**Comment by jpaddison3** [comment]
> I was imagining doing something like the latter

I just checked and there isn't unique id on raw, so that's out.

> transactions

[Transactions](https://www.postgresql.org/docs/current/tutorial-transactions.html). I was assuming you meant like a regular data migration, not a migration from self-hosted to redis. Transactions would not help with that.

**Comment by oetherington** [comment]
This solution seems reasonable to me given the contraints above.

Is there any reason why we couldn't do this with some kind of pre-built replication solution though (something like [pgpool](https://www.pgpool.net/mediawiki/index.php/Main_Page))? This should (in theory) require no code changes - we'd just setup the replication pool backed by the existing database and a new replica database, then change the connection string to the pool instead of the raw database, then copy over all the old data from before the pool existed.

File: packages/lesswrong/server/analyticsWriter.ts [hunk 2]
```diff
@@ -98,7 +99,10 @@ async function writeEventsToAnalyticsDB(events: {type, timestamp, props}[]) {
       
       inFlightRequestCounter.inFlightRequests++;
       try {
-        await connection.none(query);
+        await Promise.all([
+          connection.none(query),
```
Comment (oetherington) [inline packages/lesswrong/server/analyticsWriter.ts:103]: It looks like both `getAnalyticsConnection` and `getMirrorAnalyticsConnection` return `AnalyticsConnectionPool | null`, so we should probably either have `connection?.none(query)` checking for null (or `connection!.none(query)` if we're _really_ sure) or change the return type of `getAnalyticsConnection` to just `AnalyticsConnectionPool` if we can guaratee that.

Comment (darkruby501) [reply packages/lesswrong/server/analyticsWriter.ts:103]: Yeah, `connection?.none(query)` makes sense

**Comment by b0b3rt** [comment]
@Stvad

**Pull request merged**